### PR TITLE
Lecture 06 (janicey / Janice Yang)

### PR DESCRIPTION
**Pull request by janicey27**

### CODE REVIEW

File: notes/lecture-06-janicey.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+### Weight Quantization
+
+#### Per-Tensor vs Per-Channel Quantization
+- There often exist large differences in weights for different channels (which can lead to significant accuracy drops in smaller models). There could be large differences in ranges of weights for different output channels, or outlier weights that make all remaining weights less precise after quantization. 
+- Solution: we can use per channel quantization
+
+![Per-channel versus Per-tensor quantization example](figures/lecture-06/janicey/fig1.png)
+The figure above comes from [[Han, 2022]](https://hanlab.mit.edu/files/course/slides/MIT-TinyML-Lec06-Quantization-II.pdf), demonstrating how reconstructed error can be much smaller if we use per-channel quantization, which can lead to better model accuracy.
+
+- However, not all hardware supports this, and there might be unnecessary overhead. These days, per channel quantization is quite popular in real life applications.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:25]:
```suggestion
However, not all hardware supports this, and there might be unnecessary overhead. These days, per-channel quantization is quite popular in real-life applications. 
```

File: notes/lecture-06-janicey.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+
+## Post-Training Quantization
+How should we get the optimal linear quantization parameters (S, Z)? 
+
+### Weight Quantization
+
+#### Per-Tensor vs Per-Channel Quantization
+- There often exist large differences in weights for different channels (which can lead to significant accuracy drops in smaller models). There could be large differences in ranges of weights for different output channels, or outlier weights that make all remaining weights less precise after quantization. 
+- Solution: we can use per channel quantization
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:20]:
```suggestion
- Solution: we can use per-channel quantization
```

File: notes/lecture-06-janicey.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+![Per-channel versus Per-tensor quantization example](figures/lecture-06/janicey/fig1.png)
+The figure above comes from [[Han, 2022]](https://hanlab.mit.edu/files/course/slides/MIT-TinyML-Lec06-Quantization-II.pdf), demonstrating how reconstructed error can be much smaller if we use per-channel quantization, which can lead to better model accuracy.
+
+- However, not all hardware supports this, and there might be unnecessary overhead. These days, per channel quantization is quite popular in real life applications. 
+
+#### Weight Equalization
+- Key idea: rescaling weights to equalize the weight ranges of different output channels, so that we can potentially still use per-tensor weight quantization. 
+
+![Visualization of weight distribution and weight equalization objective](figures/lecture-06/janicey/fig2.png)
+The figure above comes from [[Markus et al., ICCV 2019]](https://hanlab.mit.edu/files/course/slides/MIT-TinyML-Lec06-Quantization-II.pdf), demonstrating how different output channel weights can be, and how they could be equalized.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:31]:
```suggestion

The figure above comes from [[Markus et al., ICCV 2019]](https://hanlab.mit.edu/files/course/slides/MIT-TinyML-Lec06-Quantization-II.pdf), demonstrating how different output channel weights can be and how they could be equalized.
```

**Review by zhijian-liu — COMMENTED** [review]
Thanks for your contribution! The scribe looks great. Please
* Fix the minor issues listed below.
* Remove all `.DS_Store` files.

File: notes/lecture-06-janicey.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+### Binary / Ternary Quantization
+- Can we push the quantization precision to just 1 bit? 
+- Changing to binary weights means we don't have to do multiplication anymore, just addition and subtraction.
+
+Types of Binarization:
+- Deterministic Binarization: directly computes the bit value based on a threshold, usually 0, resulting in a sign function. 
+- Stochastic Binarization: probability of -1 or +1 is determined by some function using global statistics or the value of input data. 
+    - e.g. in Binary Connect(BC), probability is determined by a hard sigmoid function. 
+    - this becomes harder to impelement as it requires the hardware to generate random bits when quantizing
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:118]:
```suggestion
    - this becomes harder to implement as it requires the hardware to generate random bits when quantizing
```

File: notes/lecture-06-janicey.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+### Bias Quantization
+- Key idea: fix the biased error introduced by quantization, because it can often be unevenly distributed. 
+- Method: just absorb the quantization error into the model bias parameter. 
+
+Quantization error can be denoted as $\epsilon = Q(W) - W$
+
+$E[y] = E[Wx] + E[\epsilon x] - E[\epsilon x]$
+
+$ = E[Q(W)x] - \epsilon E[x]$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:82]: Please double-check this formula

File: notes/lecture-06-janicey.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+- Type 1: During training, use Exponential moving averages (EMA). During training time, find $r_{min}$ and $r_{max}$, and smooth across thousands of training steps. 
+- Type 2: Running a few "calibration" batches of samples on the trained FP32 model. Minimize the mean-squared-error between inputs X and (reconstructed) quantized inputs Q(X), and find optimal clipping values for both ends. For Laplace (0, b) distributions, optimal clipping values can be solved numerically. Otherwise, another way to find the boundaries is to minimize Kullback-Leibler (KL) divergence, which represents the loss of information. 
+
+For two discrete probability distributions $P$ and $Q$: 
+
+$D_{KL}(P||Q) = \sum^{N}_{i}P(x_i)log\frac{P(x_i)}{Q(x_i)}$
+
+![Visualization of weight clipping after using KL Divergence optimization](figures/lecture-06/janicey/fig3.png)
+The figure above comes from [[Szymon Migacz, 2017]](https://on-demand.gputechconf.com/gtc/2017/presentation/s7310-8-bit-inference-with-tensorrt.pdf), showing an example of clipping weights on certain layer of resnet-152, after finding the optimal clipping point using KL Divergence optimization.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:72]:
```suggestion
The figure above comes from [[Szymon Migacz, 2017]](https://on-demand.gputechconf.com/gtc/2017/presentation/s7310-8-bit-inference-with-tensorrt.pdf), showing an example of clipping weights on a certain layer of ResNet-152, after finding the optimal clipping point using KL Divergence optimization.
```

File: notes/lecture-06-janicey.md (hunk 5 above)
Comment (janicey27) [reply notes/lecture-06-janicey.md:83]: This looks the same as the equations on the slides. Could you let me know which part you wanted me to change?

**Review by zhijian-liu — CHANGES_REQUESTED** [review]
The `.DS_Store` files are still there. Please remove them. Thanks!

File: notes/lecture-06-janicey.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+$ = f(W^{(i+1)}*f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
+
+$ = f(W^{(i+1)}S*f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear. 
+
+To make weight ranges of different output channels match as closely as possible, we choose $s_j$ where $s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{(r^{(i)}_{oc=j}*r^{(i+1)}_{ic=j})}$, where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$. 
+
+- Intuition, $1*100 = 10*10$, so taking product of the two numbers then the square root should give you the same product, you just get two equalized input numbers now.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:48]:
```suggestion
- Intuition, $1 * 100 = 10 * 10$, so taking the product of the two numbers, then the square root should give you the same product, and you just get two equalized input numbers now. 
```

File: notes/lecture-06-janicey.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+The figure above comes from [[Markus et al., ICCV 2019]](https://hanlab.mit.edu/files/course/slides/MIT-TinyML-Lec06-Quantization-II.pdf), demonstrating how different output channel weights can be and how they could be equalized.
+
+- Pick a scaling factor $s$, where for each $s_i$ per layer, we divide layer $i$ by s, and scale up layer $i+1$ by s. This should maintain equivariance. 
+
+Actual equations:
+
+$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)})$ 
+
+$ = f(W^{(i+1)}*f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:40]:
```suggestion
$ = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
```

File: notes/lecture-06-janicey.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+$ = f(W^{(i+1)}S*f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:42]:
```suggestion
$ = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
```

File: notes/lecture-06-janicey.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear. 
+
+To make weight ranges of different output channels match as closely as possible, we choose $s_j$ where $s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{(r^{(i)}_{oc=j}*r^{(i+1)}_{ic=j})}$, where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:46]:
```suggestion
To make weight ranges of different output channels match as closely as possible, we choose $s_j$ where $s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{(r^{(i)}_{oc=j} * r^{(i+1)}_{ic=j})}$, where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$. 
```

File: notes/lecture-06-janicey.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)})$ 
+
+$ = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
+
+$ = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear. 
+
+To make weight ranges of different output channels match as closely as possible, we choose $s_j$ where $s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{(r^{(i)}_{oc=j} * r^{(i+1)}_{ic=j})}$, where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:46]:
```suggestion
To make weight ranges of different output channels match as closely as possible, we choose $s_j$ where $s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{(r^{(i)}_{oc=j} \cdot r^{(i+1)}_{ic=j})}$, where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$. 
```

File: notes/lecture-06-janicey.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)})$ 
+
+$ = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
+
+$ = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear. 
+
+To make weight ranges of different output channels match as closely as possible, we choose $s_j$ where $s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{(r^{(i)}_{oc=j} \cdot r^{(i+1)}_{ic=j})}$, where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:46]:
```suggestion
To make the weight ranges of different output channels match as closely as possible, we choose $s_j$ where
$$s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{r^{(i)}_{oc=j} \cdot r^{(i+1)}_{ic=j}},$$
where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$. 
```

File: notes/lecture-06-janicey.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)})$ 
+
+$ = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
+
+$ = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear. 
+
+To make the weight ranges of different output channels match as closely as possible, we choose $s_j$ where
+$$s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{r^{(i)}_{oc=j} \cdot r^{(i+1)}_{ic=j}},$$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:47]:
```suggestion
$$s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{r^{(i)}_{oc=j} \cdot r^{(i+1)}_{ic=j}},$$

```

File: notes/lecture-06-janicey.md [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)})$ 
+
+$ = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
+
+$ = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear. 
+
+To make the weight ranges of different output channels match as closely as possible, we choose $s_j$ where
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:46]:
```suggestion
To make the weight ranges of different output channels match as closely as possible, we choose $s_j$ where

```

File: notes/lecture-06-janicey.md [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+$$s_j = \frac{1}{r^{(i+1)}_{ic=j}} \sqrt{r^{(i)}_{oc=j} \cdot r^{(i+1)}_{ic=j}},$$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:48]:
```suggestion
$$s_j = \frac{1}{r_{ic=j}^{(i+1)}} \sqrt{r_{oc=j}^{(i)} \cdot r_{ic=j}^{(i+1)}},$$
```

File: notes/lecture-06-janicey.md [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+$ = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear. 
+
+To make the weight ranges of different output channels match as closely as possible, we choose $s_j$ where
+
+$$s_j = \frac{1}{r_{ic=j}^{(i+1)}} \sqrt{r_{oc=j}^{(i)} \cdot r_{ic=j}^{(i+1)}},$$
+
+where $r^{(i)}_{oc=j}$ is the weight range of output channel $j$ in Layer $i$, and $r^{(i+1)}_{ic=j}$ is the weight range of input channel $j$ in layer $i+1$.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:50]:
```suggestion
where $r_{oc=j}^{(i)}$ is the weight range of output channel $j$ in Layer $i$, and $r_{ic=j}^{(i+1)}$ is the weight range of input channel $j$ in layer $i+1$. 
```

File: notes/lecture-06-janicey.md [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+where $x$ is the input to the layer, $V$ is a random variable of the same shape, $h()$ is a function to map the range to (0, 1), and $f_{reg}(V)$ is a regularization that encourages $h(V)$ to be binary. 
+
+### Activation Quantization
+- Before deploying the model, gather input activation values using some calibration batches on the trained model to understand sample distributions
+- Type 1: During training, use Exponential moving averages (EMA). During training time, find $r_{min}$ and $r_{max}$, and smooth across thousands of training steps. 
+- Type 2: Running a few "calibration" batches of samples on the trained FP32 model. Minimize the mean-squared-error between inputs X and (reconstructed) quantized inputs Q(X), and find optimal clipping values for both ends. For Laplace (0, b) distributions, optimal clipping values can be solved numerically. Otherwise, another way to find the boundaries is to minimize Kullback-Leibler (KL) divergence, which represents the loss of information. 
+
+For two discrete probability distributions $P$ and $Q$: 
+
+$D_{KL}(P||Q) = \sum^{N}_{i}P(x_i)log\frac{P(x_i)}{Q(x_i)}$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:74]:
```suggestion
$D_{KL}(P||Q) = \sum_{i}^{N}P(x_i)log\frac{P(x_i)}{Q(x_i)}$
```

File: notes/lecture-06-janicey.md [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+### Bias Quantization
+- Key idea: fix the biased error introduced by quantization, because it can often be unevenly distributed. 
+- Method: just absorb the quantization error into the model bias parameter. 
+
+Quantization error can be denoted as $\epsilon = Q(W) - W$
+
+$E[y] = E[Wx] + E[\epsilon x] - E[\epsilon x]$
+
+$ = E[Q(W)x] - \epsilon E[x]$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:87]:
```suggestion
$$E[y] = E[Wx] + E[\epsilon x] - E[\epsilon x] = E[Q(W)x] - \epsilon E[x],$$
```

File: notes/lecture-06-janicey.md [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+
+- Pick a scaling factor $s$, where for each $s_i$ per layer, we divide layer $i$ by s, and scale up layer $i+1$ by s. This should maintain equivariance. 
+
+Actual equations:
+
+$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)})$ 
+
+$ = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)})$
+
+$ = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:42]:
```suggestion
$$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)}) = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)}) = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$$
```

File: notes/lecture-06-janicey.md [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+
+The figure above comes from [[Markus et al., ICCV 2019]](https://hanlab.mit.edu/files/course/slides/MIT-TinyML-Lec06-Quantization-II.pdf), demonstrating how different output channel weights can be and how they could be equalized.
+
+- Pick a scaling factor $s$, where for each $s_i$ per layer, we divide layer $i$ by s, and scale up layer $i+1$ by s. This should maintain equivariance. 
+
+Actual equations:
+
+$$y^{(i+1)} = f(W^{(i+1)}x^{(i+1)} + b^{(i+1)}) = f(W^{(i+1)} * f(W^{(i)}x^{(i)} + b^{(i)}) + b^{(i+1)}) = f(W^{(i+1)}S * f(S^{-1}W^{(i)}x^{(i)} + S^{-1}b^{(i)}) + b^{(i+1)})$$
+
+where $S = diag(s)$, $s_j$ is the weight equalization scale factor of output channel $j$. $f$ has to be linear.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:40]:
```suggestion
where $S = \text{diag}(s)$ and $s_j$ is the weight equalization scale factor of output channel $j$. Here, $f$ has to be linear. 
```

File: notes/lecture-06-janicey.md [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+- Shortfalls: assumes simple activation functions. Works well with ReLU which is piece-weise linear, but perhaps will not work with more complex functions. 
+- Pros: does not require much computation or extra training, can get the weights and equalize them directly.
+
+#### Adaptive Rounding
+- Key idea: rounding to nearest number is not always optimal. The best rounding for each individual weight may not be the best rounding for the whole tensor, because weights interact with each other. 
+- Learn the rounding that reconstructs the original activation the best. 
+- We try to learn a parameter $delta$ between [0, 1] that optimizes rounding for each weight. 
+- Optimize the difference between the original $Wx$ and the reconstructed $\lfloor\lfloor W \rfloor + h(V)\rceil x$
+
+$argmin_v || Wx - \lfloor\lfloor W \rfloor + h(V)\rceil x ||^2_F + \lambda f_{reg}(V)$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:59]:
```suggestion
- Optimize the difference between the original $Wx$ and the reconstructed $\lfloor\lfloor W \rfloor + h(V)\rceil x$:

$$\text{argmin}_v || Wx - \lfloor\lfloor W \rfloor + h(V)\rceil x ||^2_F + \lambda f_{reg}(V)$$
```

File: notes/lecture-06-janicey.md [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+where $x$ is the input to the layer, $V$ is a random variable of the same shape, $h()$ is a function to map the range to (0, 1), and $f_{reg}(V)$ is a regularization that encourages $h(V)$ to be binary. 
+
+### Activation Quantization
+- Before deploying the model, gather input activation values using some calibration batches on the trained model to understand sample distributions
+- Type 1: During training, use Exponential moving averages (EMA). During training time, find $r_{min}$ and $r_{max}$, and smooth across thousands of training steps. 
+- Type 2: Running a few "calibration" batches of samples on the trained FP32 model. Minimize the mean-squared-error between inputs X and (reconstructed) quantized inputs Q(X), and find optimal clipping values for both ends. For Laplace (0, b) distributions, optimal clipping values can be solved numerically. Otherwise, another way to find the boundaries is to minimize Kullback-Leibler (KL) divergence, which represents the loss of information. 
+
+For two discrete probability distributions $P$ and $Q$: 
+
+$D_{KL}(P||Q) = \sum_{i}^{N}P(x_i)log\frac{P(x_i)}{Q(x_i)}$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:70]:
```suggestion
$$D_{KL}(P||Q) = \sum_{i}^{N}P(x_i)log\frac{P(x_i)}{Q(x_i)}$$
```

File: notes/lecture-06-janicey.md [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+- Shortfalls: assumes simple activation functions. Works well with ReLU which is piece-weise linear, but perhaps will not work with more complex functions. 
+- Pros: does not require much computation or extra training, can get the weights and equalize them directly.
+
+#### Adaptive Rounding
+- Key idea: rounding to nearest number is not always optimal. The best rounding for each individual weight may not be the best rounding for the whole tensor, because weights interact with each other. 
+- Learn the rounding that reconstructs the original activation the best. 
+- We try to learn a parameter $delta$ between [0, 1] that optimizes rounding for each weight. 
+- Optimize the difference between the original $Wx$ and the reconstructed $\lfloor\lfloor W \rfloor + h(V)\rceil x$:
+
+$$\text{argmin}_v || Wx - \lfloor\lfloor W \rfloor + h(V)\rceil x ||^2_F + \lambda f_{reg}(V)$$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:59]:
```suggestion
$$\arg\min_v || Wx - \lfloor\lfloor W \rfloor + h(V)\rceil x ||^2_F + \lambda f_{reg}(V)$$
```

File: notes/lecture-06-janicey.md [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+
+- Intuition, $1 * 100 = 10 * 10$, so taking the product of the two numbers, then the square root should give you the same product, and you just get two equalized input numbers now. 
+
+- Shortfalls: assumes simple activation functions. Works well with ReLU which is piece-weise linear, but perhaps will not work with more complex functions. 
+- Pros: does not require much computation or extra training, can get the weights and equalize them directly.
+
+#### Adaptive Rounding
+- Key idea: rounding to nearest number is not always optimal. The best rounding for each individual weight may not be the best rounding for the whole tensor, because weights interact with each other. 
+- Learn the rounding that reconstructs the original activation the best. 
+- We try to learn a parameter $delta$ between [0, 1] that optimizes rounding for each weight.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:56]:
```suggestion
- We try to learn a parameter $\Delta$ between [0, 1] that optimizes rounding for each weight. 
```

File: notes/lecture-06-janicey.md [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+    - this becomes harder to implement as it requires the hardware to generate random bits when quantizing
+
+Minimizing Quantization Error in Binarization 
+- Can scale by a constant value. Binary Weight Network (BWN) does this and performs significantly better than BinaryConnect. 
+
+If both activations and weights are binarized: 
+- Can be represented by a $xnor$ operation, multiplied by 2, then added to the number of terms. This involves a popcount equation and other simple operations, which is easy to implement on hardware. 
+
+Ternary Weight Networks (TWN)
+- Weights are quantized to +1, -1, and 0, using $delta = 0.7 * E(r)$ as a threshold to map out 3 distinct regions: $r > \Delta$, $|r| \leq \Delta$, and $r < -\Delta$.
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:126]:
```suggestion
- Weights are quantized to +1, -1, and 0, using $\Delta = 0.7 * E(r)$ as a threshold to map out 3 distinct regions: $r > \Delta$, $|r| \leq \Delta$, and $r < -\Delta$. 
```

File: notes/lecture-06-janicey.md [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+where $x$ is the input to the layer, $V$ is a random variable of the same shape, $h()$ is a function to map the range to (0, 1), and $f_{reg}(V)$ is a regularization that encourages $h(V)$ to be binary. 
+
+### Activation Quantization
+- Before deploying the model, gather input activation values using some calibration batches on the trained model to understand sample distributions
+- Type 1: During training, use Exponential moving averages (EMA). During training time, find $r_{min}$ and $r_{max}$, and smooth across thousands of training steps. 
+- Type 2: Running a few "calibration" batches of samples on the trained FP32 model. Minimize the mean-squared-error between inputs X and (reconstructed) quantized inputs Q(X), and find optimal clipping values for both ends. For Laplace (0, b) distributions, optimal clipping values can be solved numerically. Otherwise, another way to find the boundaries is to minimize Kullback-Leibler (KL) divergence, which represents the loss of information. 
+
+For two discrete probability distributions $P$ and $Q$: 
+
+$$D_{KL}(P||Q) = \sum_{i}^{N}P(x_i)log\frac{P(x_i)}{Q(x_i)}$$
```
Comment (zhijian-liu) [inline notes/lecture-06-janicey.md:70]:
```suggestion
$$D_{KL}(P||Q) = \sum_{i}^{N}P(x_i)\log\frac{P(x_i)}{Q(x_i)}$$
```

**Comment by zhijian-liu** [comment]
I have made the edits. Thanks again for your contribution!

**Pull request merged**